source/common/access_log/access_log_impl.cc

```diff
@@ -190,13 +199,42 @@ ResponseFlagFilter::ResponseFlagFilter(
   }
 }
 
-bool ResponseFlagFilter::evaluate(const StreamInfo::StreamInfo& info, const Http::HeaderMap&) {
+bool ResponseFlagFilter::evaluate(const StreamInfo::StreamInfo& info, const Http::HeaderMap&,
+                                  const Http::HeaderMap&) {
   if (configured_flags_ != 0) {
     return info.intersectResponseFlags(configured_flags_);
   }
   return info.hasAnyResponseFlag();
 }
 
+GrpcStatusFilter::GrpcStatusFilter(
+    const envoy::config::filter::accesslog::v2::GrpcStatusFilter& config) {
+  for (int i = 0; i < config.statuses_size(); i++) {
+    absl::optional<Grpc::Status::GrpcStatus> grpc_status =
+        Grpc::Utility::nameToGrpcStatus(config.statuses(i));
+    // Like with ResponseFlagFilter, the grpc_status will be valid because of config validation.
+    ASSERT(grpc_status.has_value());
+    statuses_.insert(grpc_status.value());
+  }
+}
+
+bool GrpcStatusFilter::evaluate(const StreamInfo::StreamInfo& info, const Http::HeaderMap&,
+                                const Http::HeaderMap& response_trailers) {
+  // The gRPC spec does not guarantee that a gRPC status will be returned from a gRPC request, so
+  // if there is no grpc-status, we must construct one from the HTTP response code.
+  const Http::HeaderEntry* entry = response_trailers.GrpcStatus();
```

Comment:
use Grpc::Common::getGrpcStatus, also note that grpc-status might appear in response_headers, when it is trailer-only response. Envoy will treat it header-only, not trailer-only.

Reply:
Thanks for the tip on `getGrpcStatus`. Still very new to the codebase!

Just to make sure I understand the trailer-only response correctly, when an HTTP response only has trailers and no header, the trailers will appear in `response_headers` instead of `response_trailers`, correct?

So order of precedence would be:

1. gRPC code from `response_trailers`, if it exists.
2. gRPC code from `response_headers`, if it exists.
3. Inferred from HTTP code from `stream_info`, if it exists.
4. `INVALID_CODE` or `UNKNOWN` (which do you think is more correct?)

Reply:
I already have testing in place for an `UNKNOWN` response, so I'll keep that as-is.